### PR TITLE
Disallow the planet label to jump around under different zooms

### LINKED ISSUE #7313: PlanetLabel scaling issue
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Planet labels change their position when scaled, even if they are not obstructed. I'm also not sure if the rings scale properly, I don't know what the desired behaviour is there.

### Steps to Reproduce

Fly around and change your zoom. Doesn't seem to happen with every planet, but when it does, it is reliable.

### Expected Behavior

The text shouldn't move.

### Screenshots

https://user-images.githubusercontent.com/68112292/193869561-e5619a64-a4b9-41e8-b2f9-4e6be5b21a5d.mp4

### Link to save file

_No response_

### Operating System

EndeavourOS (rolling), Wayland, Gnome 42.4

### Game Source

ESLauncher2

### Game Version

v0.9.15

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by quyykk**
**Bugfix:** This PR addresses fixes #7313 

## Fix Details

This PR does two things:

1. Considers the planet's outer ring when checking for overlap with a label.
2. Prefers a direction where under any zoom level the labels stay in the same direction. This is to avoid the labels jumping around when zooming in/out.

## Testing Done

Used the scenario described in the linked issue. Without this PR the label jumps around, with this PR it stays on the left side.

## Performance Impact

It does a couple more overlap checks per frame (8 to be exact). But I don't expect those to impact performance in any meaningful way.

### CODE REVIEW

**Review by tibetiroka — APPROVED** [review]
Comment grammar looks good. I'm not awake enough to check the actual code.

File: source/PlanetLabel.cpp [hunk 1]
```diff
@@ -39,6 +40,41 @@ namespace {
...
 	const double GAP = 6.;
 	const double MIN_DISTANCE = 30.;
+
+	// Check if the given label for the given stellar object and direction overlaps with any other stellar object in the system.
+	bool Overlaps(const System &system, const StellarObject &object, double zoom, double width, int direction)
+	{
+		bool overlaps = false;
+
+		Point start = object.Position() * zoom +
+			(object.Radius() + INNER_SPACE + LINE_GAP + LINE_LENGTH) * zoom * Angle(LINE_ANGLE[direction]).Unit();
```
Comment (Amazinite) [inline source/PlanetLabel.cpp:50]: Can't this be simplified now?
```suggestion
		Point start = zoom * (object.Position() +
			(object.Radius() + INNER_SPACE + LINE_GAP + LINE_LENGTH) * Angle(LINE_ANGLE[direction]).Unit());
```

File: source/PlanetLabel.cpp [hunk 2]
```diff
@@ -39,6 +40,41 @@ namespace {
...
+		{
+			if(&other == &object)
+				continue;
+
+			double minDistance = (other.Radius() + MIN_DISTANCE) * zoom;
+
+			Point otherPos = other.Position() * zoom;
+			double startDistance = otherPos.Distance(start);
+			double endDistance = otherPos.Distance(end);
+			overlaps |= (startDistance < minDistance || endDistance < minDistance);
```
Comment (petervdmeer) [inline source/PlanetLabel.cpp:64]: If an overlap is found, then this function can immediately return true.

File: source/PlanetLabel.cpp [hunk 3]
```diff
@@ -39,6 +40,41 @@ namespace {
...
+			double projection = (otherPos - start).Dot(unit);
+
+			if(projection > 0. && projection < width)
```
Comment (petervdmeer) [inline source/PlanetLabel.cpp:67]: This code appears to check for "overlap with the middle of the label, when the end and/or start might not overlap". Maybe good to add a comment above this code to describe this for the maintainers that need to deal with this code in the future.

File: source/PlanetLabel.cpp [hunk 4]
```diff
@@ -39,6 +40,41 @@ namespace {
...
 	const double MIN_DISTANCE = 30.;
+
+	// Check if the given label for the given stellar object and direction overlaps with any other stellar object in the system.
+	bool Overlaps(const System &system, const StellarObject &object, double zoom, double width, int direction)
+	{
+		bool overlaps = false;
+
+		Point start = zoom * (object.Position() +
+			(object.Radius() + INNER_SPACE + LINE_GAP + LINE_LENGTH) * Angle(LINE_ANGLE[direction]).Unit());
+		Point unit(LINE_ANGLE[direction] > 180. ? -1. : 1., 0.);
```
Comment (petervdmeer) [inline source/PlanetLabel.cpp:51]: This converts from "all possible directions" to "only leftward and rightward facing vectors". For the labels that is currently fine, but maybe add a comment here to describe that this conversion is fine for this type of label.

**Review by petervdmeer — APPROVED** [review]
Code looks good. I have some minor comments, but I don't need to re-review when those are handled (or ignored if you disagree with the comments).

**Pull request merged**